### PR TITLE
Improve the handling of map options

### PR DESCRIPTION
**Pull request by vicb**
Content of this PR:

- Fix comments in `type MapOptions`
- Order default values using the type order,
- Update Map props:
  - use inline init where applicable
  - fix types where applicable
  - fix init to match type where applicable
- A few other minor updates

I'd like to get some feedback and I'll create an issue and fill out the launch checklist after that.

Thanks! 

## Launch Checklist

 - [ ] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [ ] Briefly describe the changes in this PR.
 - [ ] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: src/ui/map.ts [hunk 1]
```diff
@@ -553,80 +552,76 @@ export class Map extends Camera {
...
+        this._refreshExpiredTiles = resolvedOptions.refreshExpiredTiles === true;
+        this._fadeDuration = resolvedOptions.fadeDuration;
+        this._crossSourceCollisions = resolvedOptions.crossSourceCollisions === true;
+        this._collectResourceTiming = resolvedOptions.collectResourceTiming === true;
+        this._locale = {...defaultLocale, ...resolvedOptions.locale};
+        this._clickTolerance = resolvedOptions.clickTolerance;
+        this._overridePixelRatio = resolvedOptions.pixelRatio;
+        this._maxCanvasSize = resolvedOptions.maxCanvasSize;
+        this.transformCameraUpdate = resolvedOptions.transformCameraUpdate;
+        this.cancelPendingTileRequestsWhileZooming = resolvedOptions.cancelPendingTileRequestsWhileZooming === true;
```
Comment (HarelM) [inline src/ui/map.ts:593]: Why this `=== true` is needed?

Comment (HarelM) [reply src/ui/map.ts:593]: I see it in a few more lines.

Comment (vicb) [reply src/ui/map.ts:593]: You're right that's is not needed here.

But for i.e. `failIfMajorPerformanceCaveat?: boolean;`, `_failIfMajorPerformanceCaveat: boolean;` is not expected to be undefined so `=== true` would transofm `undefined` to `false`.

I'll update to remove the place where it's not required.

File: src/types/util.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ *  bar: number | undefined;
+ *  baz: number | undefined;
+ * }
+ *
+ * ```
+ *
+ * See https://medium.com/terria/typescript-transforming-optional-properties-to-required-properties-that-may-be-undefined-7482cb4e1585
+ */
+
+export type Complete<T> = {
```
Comment (HarelM) [inline src/types/util.ts:22]: I was actually thinking of removing this folder entirely, and move the types to either the utils folder or the files that are using those.
I also don't see this is used in another place, so I'm not sure there's an actual need to move it...

**Comment by HarelM** [comment]
I don't have any objections to this PR in general.
Can we close the followin PR:
- #4003?

File: src/types/util.ts (hunk 2 above)
Comment (vicb) [reply src/types/util.ts:22]: I prefer not to have too much noise in `map.ts` but no strong opinion, I can do whatever you think is best.

One other unrelated question, do you know why `const defaultMinZoom = -2;`?
I would love to add a comment there because it doesn't look intuitive

**Comment by vicb** [comment]
> I don't have any objections to this PR in general. Can we close the following PR:
> 
> * [Fix MapOptions#style is optional in the Map constructor #4003](https://github.com/maplibre/maplibre-gl-js/pull/4003)?

For now `style` is not addressed in this PR but I think it should be.
Let's close the other PR once this has been updated.
The first commit is only about cleaning the existing stuff. I'll add more commits to improve the type after we agree on the first one.

File: src/types/util.ts (hunk 2 above)
Comment (vicb) [reply src/types/util.ts:22]: -2 is to render maps < 512px (see https://github.com/mapbox/mapbox-gl-js/pull/9028)

**Comment by vicb** [comment]
It looks like some tests are failing...
I'll try to reproduce locally and fix

**Comment by HarelM** [comment]
I'm not sure the failing test is your fault. I see it on a PR for a package update.
Wait a bit with the investigation...

File: src/ui/map.ts [hunk 3]
```diff
@@ -1727,28 +1721,30 @@ export class Map extends Camera {
...
-            this.style.once('style.load', () => this._updateStyle(style, options));
-            return;
-        }
-
-        const previousStyle = this.style && options.transformStyle ? this.style.serialize() : undefined;
+    _updateStyle(style: StyleSpecification | string | null | undefined, options?: StyleSwapOptions & StyleOptions) {
+        let previousStyle: StyleSpecification | undefined;
         if (this.style) {
-            this.style.setEventedParent(null);
+            // transformStyle relies on having previous style serialized.
```
Comment (HarelM) [inline src/ui/map.ts:1727]: Is this a bug fix? If so, I think it will be better to have this in a different PR...?

File: src/ui/map.ts [hunk 4]
```diff
@@ -2181,8 +2172,8 @@ export class Map extends Camera {
      * let coffeeShopIcon = map.getImage("coffee_cup");
      * ```
      */
-    getImage(id: string): StyleImage {
-        return this.style.getImage(id);
+    getImage(id: string): StyleImage | undefined {
+        return this.style?.getImage(id);
```
Comment (HarelM) [inline src/ui/map.ts:2176]: Isn't this a behavior change? As far as I understand, before this change if style was not defined, this method would have thrown and exception, wouldn't it?

**Comment by HarelM** [comment]
I believe this PR got too big... :-(
Let's try and keep the PRs small, with small increments so these increments can be merged.
From previous commit to latest commit there are a lot of changes which I think should be discussed before merging.
The previous commit was great in terms of adding correctness to types without changing any implementation (maybe besides moving the initialization out of the constructor).
Let's start with that and then move on to fixing things/changing logic.

File: src/ui/map.ts [hunk 5]
```diff
@@ -298,14 +301,15 @@ export type MapOptions = {
      * the schema described in the [MapLibre Style Specification](https://maplibre.org/maplibre-style-spec/),
      * or a URL to such JSON.
      */
-    style: StyleSpecification | string;
+    style?: StyleSpecification | string;
```
Comment (HarelM) [inline src/ui/map.ts:304]: I think that other improvements are great, I would avoid changing this as part of this PR, keep it as is for now and open a different PR just for this change (with its impact obviously).

File: src/ui/map.ts (hunk 3 above)
Comment (vicb) [reply src/ui/map.ts:1727]: This is only code factorization:
- `this.style` is tested on l 1732, 1737, and 1738 on the left
- `options.transformStyle` is tested l 1732 and 1737 on the left

File: src/ui/map.ts (hunk 4 above)
Comment (vicb) [reply src/ui/map.ts:2176]: Discussed offline with @HarelM 

Summary of the discussion: What is the expected behavior of this function:

- throw is not documented
- the type of `this.style: Style` does not let users think it might throw
- is it valuable to throw here?

I think the behavior should be consistent across methods.

Some might throw:

```ts
    getImage(id: string): StyleImage {
        return this.style.getImage(id);
    }
```

Some are noops when `style` is undefined

```ts
    redraw(): this {
        if (this.style) {
            // cancel the scheduled update
            if (this._frameRequest) {
                this._frameRequest.abort();
                this._frameRequest = null;
            }
            this._render(0);
        }
        return this;
    }
```

Some warn when `style` is undefined

```ts
   _updateDiff(style: StyleSpecification, options?: StyleSwapOptions & StyleOptions) {
        try {
            if (this.style.setState(style, options)) {
                this._update(true);
            }
        } catch (e) {
            warnOnce(
                `Unable to perform style diff: ${e.message || e.error || e}.  Rebuilding the style from scratch.`
            );
            this._updateStyle(style, options);
        }
    }
```

**Review by vicb — COMMENTED** [review]
@HarelM 

Maybe the best way to move forward is to remove the last commit from this PR.

I can then submit another PR making `MapOptions.style` optional and dropping all the other changes in that last commit as I don't think you find them useful?

**Comment by HarelM** [comment]
Yes, let's drop the last commit and open an issue about the inconsistency of the methods.
Once we agree on the way forward in the issue we can continue to open a PR with the agreed upon changes.
It's not that I don't think they are valuable, it's just that they need a bit more discussion as some of them are breaking changes to public API methods in terms of behavior and I would like to make sure we discuss them properly.
Generally speaking, I agree that the current API is not consistent, and we should dive into it and see what improvements are considered low hanging fruits and what will need a breaking change as part of a breaking change version (see #3834) , but I believe it is out of scope for this PR.
In general, I would prefer to better document the current behavior in types and method docs as an initial non-breaking-change manner, and then later on, if needed, change the behavior as part of a breaking change version.

**Review by HarelM — APPROVED** [review]
THANKS!

**Comment by vicb** [comment]
Thanks for your help Harel!
